test/annotation_utils.t

```diff
@@ -864,6 +980,112 @@ subtest '_run_rast_workflow_ann' => sub {
+        }
+        else {
+            print "\nThere is NO contigID_hash in $genome_before_remapping->{id}:\n".Dumper(keys %rd);
+            next;
+        }
+
+        ## Run the remapping
+        my $genome_after_remapping = $annoutil->_remap_contigIDs( $ctgID_hash, $genome_before_remapping );
+        print "\n=========after remapping the rasted genome on $genome_before_remapping->{id}:\n";
+
+        my $exp_ctg_ids = [];
+        if( $genome_after_remapping->{contig_ids} &&
+                @{$genome_after_remapping->{contig_ids}} > 0) {
+            for my $cid (@{$genome_before_remapping->{contig_ids}}) {
+                if( $ctgID_hash->{$cid} ) {
+                    push @{$exp_ctg_ids}, $ctgID_hash->{$cid};
+                } else {
+                    push @{$exp_ctg_ids}, $cid;
+                }
+            }
+            cmp_deeply $genome_after_remapping->{contig_ids}, $exp_ctg_ids,
+                        'contig ids remapped correctly';
+        }
+
+        $exp_ctg_ids = [];
+        my $result_ctg_ids = [];
+        if( $genome_after_remapping->{contigs} &&
+                @{$genome_after_remapping->{contigs}} > 0) {
+            for my $ctg_before (@{$genome_before_remapping->{contigs}}) {
+                if( $ctgID_hash->{$ctg_before->{id}} ) {
+                    push @{$exp_ctg_ids}, $ctgID_hash->{$ctg_before->{id}};
+                } else {
+                    push @{$exp_ctg_ids}, $ctg_before->{id};
+                }
+            }
+            for my $ctg_after (@{$genome_after_remapping->{contigs}}) {
+                push @{$result_ctg_ids}, $ctg_after->{id};
+            }
+            cmp_deeply $result_ctg_ids, $exp_ctg_ids,
+                        'contig ids of contigs array remapped correctly';
+        }
+
+        my $exp_locs = [];
+        my $result_locs = [];
+        if( $genome_after_remapping->{features} &&
+                @{$genome_after_remapping->{features}} > 0) {
+            for my $ftr_after (@{$genome_after_remapping->{features}}) {
+                push @{$result_locs}, $ftr_after->{location};
+            }
+            $exp_locs = get_feature_locations($ctgID_hash, $genome_before_remapping->{features});
+
+            cmp_deeply $result_locs, $exp_locs,
+                        'contig ids of feature locations remapped correctly';
+        }
+
+        $result_locs = [];
+        if( $genome_after_remapping->{non_coding_features} &&
+                @{$genome_after_remapping->{non_coding_features}} > 0) {
+            for my $ncftr_after (@{$genome_after_remapping->{non_coding_features}}) {
+                push @{$result_locs}, $ncftr_after->{location};
+            }
+            $exp_locs = get_feature_locations(
+                            $ctgID_hash, $genome_before_remapping->{non_coding_features});
+
+            cmp_deeply $result_locs, $exp_locs,
+                        'contig ids of non_coding_feature locations remapped correctly';
+        }
```

Comment:
This can still be abstracted out again as you still have repeated code blocks:

```
for my $feature_type ( qw( features non_coding_features css mrnas ) ) {
  if ( $genome_after_remapping->{ $feature_type } && @{ $genome_after_remapping->{ $feature_type } } ) {
    my @got_locations = map { $_->{ location } } @{ $genome_after_remapping->{ $feature_type } };
    ...
}
```
etc.

Reply:
No, I don't know how perl map function works and dare not to try here.